File: runtime/env.py
# Ethan Doughty
# env.py
"""Environment for tracking variable shapes during analysis."""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, Any, Optional

from runtime.shapes import Shape, join_shape, widen_shape


@dataclass
class Env:
    """Mapping from variable names to their inferred shapes.

    Supports optional parent-pointer scope chains for nested scoping.
    get() walks the chain; set() writes to local scope only.
    """
    bindings: Dict[str, Shape] = field(default_factory=dict)
    dim_aliases: Dict[str, Any] = field(default_factory=dict)
    parent: Optional[Env] = None

    def copy(self) -> Env:
        """Shallow copy of local scope, sharing the same parent."""
        return Env(bindings=self.bindings.copy(),
                   dim_aliases=self.dim_aliases.copy(),
                   parent=self.parent)

    def get(self, name: str) -> Shape:
        """Get shape by name, walking parent chain if not local."""
        if name in self.bindings:
            return self.bindings[name]
        if self.parent is not None:
            return self.parent.get(name)
        return Shape.bottom()

    def set(self, name: str, shape: Shape) -> None:
        """Set shape in local scope."""
        self.bindings[name] = shape

    def has_local(self, name: str) -> bool:
        """Check if name is bound in local scope (not parent)."""
        return name in self.bindings

    def __contains__(self, name: str) -> bool:
        """Check if name is bound anywhere in the scope chain."""
        if name in self.bindings:
            return True
        return self.parent is not None and name in self.parent

    def push_scope(self) -> Env:
        """Create a child scope with this env as parent."""
        return Env(parent=self)

    def replace_local(self, other: Env) -> None:
        """Replace local bindings and aliases from another env."""
        self.bindings = other.bindings
        self.dim_aliases = other.dim_aliases

    def local_bindings_equal(self, other: Env) -> bool:
        """Compare local bindings only (not parent chain)."""
        return self.bindings == other.bindings

    def __repr__(self) -> str:
        """String representation showing all bindings."""
        parts = [f"{var_name}: {shape}" for var_name, shape in self.bindings.items()]
        return "Env{" + ", ".join(parts) + "}"


def join_env(env1: Env, env2: Env) -> Env:
    """Merge two environments by joining shapes pointwise.

    Args:
        env1: First environment (e.g., from then-branch)
        env2: Second environment (e.g., from else-branch)

    Returns:
        Merged environment with joined shapes for each variable
    """
    result = Env(parent=env1.parent)
    all_vars = sorted(set(env1.bindings.keys()) | set(env2.bindings.keys()))
    for var_name in all_vars:
        shape1 = env1.bindings.get(var_name, Shape.bottom())
        shape2 = env2.bindings.get(var_name, Shape.bottom())
        result.set(var_name, join_shape(shape1, shape2))
    return result


def widen_env(env1: Env, env2: Env) -> Env:
    """Widen two environments pointwise. Used for both widening and post-loop join.

    Used in fixpoint loop analysis to accelerate convergence by widening
    conflicting dimensions to None (unknown). Also used for post-loop join
    to model "loop may not execute" semantics.

    Args:
        env1: First environment (old/pre-loop state)
        env2: Second environment (new/post-iteration state)

    Returns:
        Widened environment with pointwise widen_shape applied
    """
    result = Env(parent=env1.parent)
    all_vars = sorted(set(env1.bindings.keys()) | set(env2.bindings.keys()))
    for var_name in all_vars:
        shape1 = env1.bindings.get(var_name, Shape.bottom())
        shape2 = env2.bindings.get(var_name, Shape.bottom())
        result.set(var_name, widen_shape(shape1, shape2))
    return result

File: src/Env.fs
module Env

open Shapes

// ---------------------------------------------------------------------------
// Env: variable name -> Shape scope chain
// get() walks the parent chain; set() writes local scope only.
// ---------------------------------------------------------------------------

type Env = {
    mutable bindings:   Map<string, Shape>
    mutable dimAliases: Map<string, Dim>
    parent:             Env option
}

module Env =

    let create () : Env = {
        bindings   = Map.empty
        dimAliases = Map.empty
        parent     = None
    }

    let createWithParent (parent: Env) : Env = {
        bindings   = Map.empty
        dimAliases = Map.empty
        parent     = Some parent
    }

    let rec get (env: Env) (name: string) : Shape =
        match Map.tryFind name env.bindings with
        | Some s -> s
        | None ->
            match env.parent with
            | Some p -> get p name
            | None   -> Bottom

    let set (env: Env) (name: string) (shape: Shape) : unit =
        env.bindings <- Map.add name shape env.bindings

    let hasLocal (env: Env) (name: string) : bool =
        Map.containsKey name env.bindings

    let rec contains (env: Env) (name: string) : bool =
        Map.containsKey name env.bindings ||
        (match env.parent with
         | Some p -> contains p name
         | None   -> false)

    let copy (env: Env) : Env = {
        bindings   = env.bindings
        dimAliases = env.dimAliases
        parent     = env.parent
    }

    let pushScope (env: Env) : Env = createWithParent env

    let replaceLocal (env: Env) (other: Env) : unit =
        env.bindings   <- other.bindings
        env.dimAliases <- other.dimAliases

    let localBindingsEqual (env: Env) (other: Env) : bool =
        env.bindings = other.bindings

// ---------------------------------------------------------------------------
// Module-level join/widen (mirror Python's join_env / widen_env)
// ---------------------------------------------------------------------------

let joinEnv (env1: Env) (env2: Env) : Env =
    let result = {
        bindings   = Map.empty
        dimAliases = Map.empty
        parent     = env1.parent
    }
    let allVars =
        Set.union
            (env1.bindings |> Map.toSeq |> Seq.map fst |> Set.ofSeq)
            (env2.bindings |> Map.toSeq |> Seq.map fst |> Set.ofSeq)
    for var in allVars |> Set.toList |> List.sort do
        let s1 = defaultArg (Map.tryFind var env1.bindings) Bottom
        let s2 = defaultArg (Map.tryFind var env2.bindings) Bottom
        Env.set result var (joinShape s1 s2)
    result

let widenEnv (env1: Env) (env2: Env) : Env =
    let result = {
        bindings   = Map.empty
        dimAliases = Map.empty
        parent     = env1.parent
    }
    let allVars =
        Set.union
            (env1.bindings |> Map.toSeq |> Seq.map fst |> Set.ofSeq)
            (env2.bindings |> Map.toSeq |> Seq.map fst |> Set.ofSeq)
    for var in allVars |> Set.toList |> List.sort do
        let s1 = defaultArg (Map.tryFind var env1.bindings) Bottom
        let s2 = defaultArg (Map.tryFind var env2.bindings) Bottom
        Env.set result var (widenShape s1 s2)
    result
